// vite-plugin-electron/dist/base-DQFn5ze5.mjs
import { d as setupMockHtml, o as resolveInput, s as resolveServerUrl, t as checkESModule, u as setIsViteDev } from "./utils-DzpHSY0b.mjs";
import path from "node:path";
//#region src/startup.ts
const startupEnv = {
	REMOTE_DEBUGGING_PORT: "--remote-debugging-port",
	ELECTRON_IGNORE_CERTIFICATE_ERRORS: "--ignore-certificate-errors",
	ELECTRON_DISABLE_WEB_SECURITY: "--disable-web-security",
	ELECTRON_INSPECT: "--inspect",
	ELECTRON_INSPECT_BRK: "--inspect-brk"
};
function parseEnvVar(value) {
	if (value === "true" || value === "1") return true;
	if (value === "false" || value === "0" || value === "" || value === void 0) return false;
	return value;
}
/**
* Electron App startup function.
* It will mount the Electron App child-process to `process.electronApp`.
*
* You can also set environment variables to control the Electron CLI flags.
* `1` or `true` turns a flag on, `0` or `false` turns it off, and any other non-empty
* value is appended as `=<value>`.
*
* Supported env vars:
* - `REMOTE_DEBUGGING_PORT` appends `--remote-debugging-port=<value>`
* - `ELECTRON_IGNORE_CERTIFICATE_ERRORS` appends `--ignore-certificate-errors`
* - `ELECTRON_DISABLE_WEB_SECURITY` appends `--disable-web-security`
* - `ELECTRON_INSPECT` appends `--inspect` or `--inspect=<value>`
* - `ELECTRON_INSPECT_BRK` appends `--inspect-brk` or `--inspect-brk=<value>`
* @param argv default value `['.', '--no-sandbox']`
* @param options options for `child_process.spawn`
* @param customElectronPkg custom electron package name (default: 'electron')
* @returns `true` if the Electron app is started, or `false` if the startup is prevented by `startup.prevent` or `ELECTRON_STARTUP_PREVENT` env var.
*/
const startup = async (argv = [".", "--no-sandbox"], options, customElectronPkg) => {
	if (startup.prevent || parseEnvVar(process.env.ELECTRON_STARTUP_PREVENT?.trim())) {
		process.electronApp = void 0;
		return false;
	}
	const { spawn } = await import("node:child_process");
	const { createRequire } = await import("node:module");
	const electronPackage = customElectronPkg ?? "electron";
	const roots = new Set([
		...typeof options?.cwd === "string" ? [options.cwd] : [],
		...process.env.INIT_CWD ? [process.env.INIT_CWD] : [],
		process.cwd()
	]);
	let electron;
	let resolutionError;
	for (const root of roots) try {
		electron = createRequire(path.join(root, "package.json"))(electronPackage);
		break;
	} catch (error) {
		resolutionError = error;
	}
	if (!electron) try {
		electron = await import(electronPackage);
	} catch (error) {
		resolutionError = error;
	}
	if (!electron) {
		process.electronApp = void 0;
		throw new Error(`Unable to resolve "${electronPackage}". Install it in the app project or pass startup(..., ..., customElectronPkg).`, { cause: resolutionError });
	}
	const electronPath = electron.default ?? electron;
	startup.exit();
	const stdio = process.platform === "linux" ? [
		"inherit",
		"inherit",
		"inherit",
		"ignore",
		"ipc"
	] : [
		"inherit",
		"inherit",
		"inherit",
		"ipc"
	];
	const targetArgv = [...argv];
	for (const [envName, flag] of Object.entries(startupEnv)) {
		const value = parseEnvVar(process.env[envName]?.trim());
		if (!value) continue;
		if (value === true) targetArgv.push(flag);
		else targetArgv.push(`${flag}=${value}`);
	}
	process.electronApp = spawn(electronPath, targetArgv, {
		stdio,
		...options
	});
	process.electronApp.on("exit", process.exit);
	return true;
};
startup.send = (message) => {
	if (process.electronApp) process.electronApp.send?.(message);
};
startup.hookedProcessExit = startup.prevent = false;
startup.exit = () => {
	process.electronApp?.removeAllListeners();
	process.electronApp?.kill();
};
/**
* Trigger the startup of the Electron app during development.
* If `options.onstart` is provided, it will be called with a `startup` function and a `reload` function.
* Otherwise, the Electron app will start immediately.
* @param context The `this` of `configServer()` hook, used for calling `this` in `options.onstart`.
* @param server The Vite development server instance.
* @param options The `onstart` options
*/
function triggerStartup(context, server, options) {
	const startupWithRoot = (argv, spawnOptions, customElectronPkg) => {
		return startup(argv, {
			cwd: server.config.root,
			...spawnOptions
		}, customElectronPkg);
	};
	if (options.onstart) options.onstart.call(context, {
		startup: startupWithRoot,
		reload() {
			if (process.electronApp) {
				(server.hot || server.ws).send({ type: "full-reload" });
				startup.send("electron-vite&type=hot-reload");
			} else startupWithRoot();
		}
	});
	else startupWithRoot();
}
const defaultPreloadOnstart = async (args) => {
	args.reload();
};
//#endregion
//#region src/base.ts
function createElectronPlugin({ prefix, buildConfig, dev, build }) {
	let userConfig;
	let configEnv;
	let cleanupMock;
	let isESM;
	return [{
		name: `${prefix}:dev`,
		apply: "serve",
		config(_, env) {
			if (env.command === "serve") setIsViteDev();
		},
		configResolved(config) {
			isESM = checkESModule(config.root);
			if (!resolveInput(config)) cleanupMock = setupMockHtml(config, false, config.logger);
		},
		configureServer(server) {
			server.httpServer?.once("close", async () => {
				if (cleanupMock) {
					await cleanupMock();
					cleanupMock = void 0;
				}
				startup.exit();
			});
			server.httpServer?.once("listening", async () => {
				Object.assign(process.env, { VITE_DEV_SERVER_URL: resolveServerUrl(server) });
				await dev(this, server, isESM);
			});
		}
	}, {
		name: `${prefix}:prod`,
		apply: "build",
		async config(config, env) {
			userConfig = config;
			configEnv = env;
			return {
				...config.base ? {} : { base: "./" },
				...await buildConfig?.(config, env)
			};
		},
		configResolved(config) {
			isESM = checkESModule(config.root);
			if (!resolveInput(config)) cleanupMock = setupMockHtml(config, true, config.logger);
		},
		async closeBundle() {
			try {
				await build(userConfig, configEnv, isESM);
			} finally {
				if (cleanupMock) {
					await cleanupMock();
					cleanupMock = void 0;
				}
			}
		}
	}];
}
//#endregion
export { triggerStartup as i, defaultPreloadOnstart as n, startup as r, createElectronPlugin as t };
