// vite-plugin-electron/dist/index.mjs
import { i as triggerStartup, r as startup, t as createElectronPlugin } from "./base-DQFn5ze5.mjs";
import { c as resolveViteConfig, l as resolveViteConfigBase, n as compatRollupOptions, p as withExternalBuiltins, t as checkESModule } from "./utils-DzpHSY0b.mjs";
import { build as build$1 } from "vite";
import { loadPackageJSON, loadPackageJSONSync } from "local-pkg";
//#region src/index.ts
function build(options) {
	return buildBase(checkESModule(), options);
}
function buildBase(isESM, options) {
	return build$1(withExternalBuiltins(resolveViteConfigBase(isESM, options)));
}
function electron(options) {
	const optionsArray = Array.isArray(options) ? options : [options];
	return createElectronPlugin({
		prefix: "vite-plugin-electron",
		async dev(pluginContext, server, isESM) {
			const entryCount = optionsArray.length;
			let closeBundleCount = 0;
			for (const options of optionsArray) {
				options.vite ??= {};
				options.vite.mode ??= server.config.mode;
				options.vite.root ??= server.config.root;
				options.vite.envDir ??= server.config.envDir;
				options.vite.envPrefix ??= server.config.envPrefix;
				options.vite.build ??= {};
				if (!("watch" in options.vite.build)) options.vite.build.watch = {};
				options.vite.build.minify ??= false;
				options.vite.plugins ??= [];
				options.vite.plugins.push({
					name: ":startup",
					closeBundle() {
						if (++closeBundleCount < entryCount) return;
						triggerStartup(pluginContext, server, options);
					}
				});
				await buildBase(isESM, options);
			}
		},
		async build(userConfig, configEnv, isESM) {
			for (const options of optionsArray) {
				options.vite ??= {};
				options.vite.mode ??= configEnv.mode;
				options.vite.root ??= userConfig.root;
				options.vite.envDir ??= userConfig.envDir;
				options.vite.envPrefix ??= userConfig.envPrefix;
				await buildBase(isESM, options);
			}
		}
	});
}
//#endregion
export { build, checkESModule, compatRollupOptions, electron as default, loadPackageJSON, loadPackageJSONSync, resolveViteConfig, startup, withExternalBuiltins };
