// vite-plugin-electron/dist/index.cjs
Object.defineProperties(exports, {
	__esModule: { value: true },
	[Symbol.toStringTag]: { value: "Module" }
});
const require_utils = require("./utils-C-Tpxd3l.cjs");
const require_base = require("./base-BISTNoCX.cjs");
let vite = require("vite");
let local_pkg = require("local-pkg");
//#region src/index.ts
function build(options) {
	return buildBase(require_utils.checkESModule(), options);
}
function buildBase(isESM, options) {
	return (0, vite.build)(require_utils.withExternalBuiltins(require_utils.resolveViteConfigBase(isESM, options)));
}
function electron(options) {
	const optionsArray = Array.isArray(options) ? options : [options];
	return require_base.createElectronPlugin({
		prefix: "vite-plugin-electron",
		async dev(pluginContext, server, isESM) {
			const entryCount = optionsArray.length;
			let closeBundleCount = 0;
			for (const options of optionsArray) {
				options.vite ??= {};
				options.vite.mode ??= server.config.mode;
				options.vite.root ??= server.config.root;
				options.vite.envDir ??= server.config.envDir;
				options.vite.envPrefix ??= server.config.envPrefix;
				options.vite.build ??= {};
				if (!("watch" in options.vite.build)) options.vite.build.watch = {};
				options.vite.build.minify ??= false;
				options.vite.plugins ??= [];
				options.vite.plugins.push({
					name: ":startup",
					closeBundle() {
						if (++closeBundleCount < entryCount) return;
						require_base.triggerStartup(pluginContext, server, options);
					}
				});
				await buildBase(isESM, options);
			}
		},
		async build(userConfig, configEnv, isESM) {
			for (const options of optionsArray) {
				options.vite ??= {};
				options.vite.mode ??= configEnv.mode;
				options.vite.root ??= userConfig.root;
				options.vite.envDir ??= userConfig.envDir;
				options.vite.envPrefix ??= userConfig.envPrefix;
				await buildBase(isESM, options);
			}
		}
	});
}
//#endregion
exports.build = build;
exports.checkESModule = require_utils.checkESModule;
exports.compatRollupOptions = require_utils.compatRollupOptions;
exports.default = electron;
Object.defineProperty(exports, "loadPackageJSON", {
	enumerable: true,
	get: function() {
		return local_pkg.loadPackageJSON;
	}
});
Object.defineProperty(exports, "loadPackageJSONSync", {
	enumerable: true,
	get: function() {
		return local_pkg.loadPackageJSONSync;
	}
});
exports.resolveViteConfig = require_utils.resolveViteConfig;
exports.startup = require_base.startup;
exports.withExternalBuiltins = require_utils.withExternalBuiltins;
