// vite-plugin-electron/dist/multi-env.mjs
import { i as triggerStartup, n as defaultPreloadOnstart, t as createElectronPlugin } from "./base-DQFn5ze5.mjs";
import { f as toArray, i as createElectronViteDefaults, n as compatRollupOptions, p as withExternalBuiltins, r as createDefaultPreloadConfig, t as checkESModule } from "./utils-DzpHSY0b.mjs";
import { n as extractExternalDeps } from "./notBundle-JWunStGa.mjs";
import { createBuilder, mergeConfig, perEnvironmentPlugin } from "vite";
import { loadPackageJSONSync } from "local-pkg";
//#region src/multi-env.ts
/**
* Helper function to create simple API for {@link electron} like `vite-plugin-electron/simple`.
*/
function simpleOptions(options) {
	return Object.entries(options).map(([name, { options, ...rest }]) => {
		switch (name) {
			case "main": return Object.assign(rest, {
				name,
				options: mergeConfig({ build: compatRollupOptions({ rolldownOptions: { platform: "node" } }) }, options ?? {})
			});
			case "preload": return Object.assign(rest, {
				name,
				onstart: rest.onstart || defaultPreloadOnstart,
				options: mergeConfig(createDefaultPreloadConfig(checkESModule()), options ?? {})
			});
			default: return Object.assign(rest, {
				name,
				options
			});
		}
	});
}
function electronSimple(options) {
	return electron(simpleOptions(options));
}
const PLUGIN_PREFIX = "vite-plugin-electron-multi-env";
function electronPluginFactory(options) {
	const resolveOptions = async (isDev, root) => {
		const context = {
			root,
			packageJson: loadPackageJSONSync(root),
			isDev
		};
		if (!context.packageJson) throw new Error("[vite-plugin-electron] Cannot find package.json");
		const rawOptions = await options(context);
		const envNames = /* @__PURE__ */ new Set();
		const environmentOptions = toArray(rawOptions).map((opt, i) => {
			const optionName = opt.name || i;
			const name = `electron_${optionName}`;
			if (envNames.has(name)) throw new Error(`[vite-plugin-electron] Duplicate environment name "${optionName}". Please provide unique "name" properties for each environment in the options array.`);
			envNames.add(name);
			return Object.assign({}, opt, { name });
		});
		return {
			environmentOptions,
			defaultEnvs: Object.fromEntries(environmentOptions.map((opt) => {
				const defaultConfig = createElectronViteDefaults(context.packageJson.type === "module", {
					input: opt.input,
					plugins: opt.plugins,
					external: !opt.notBundle ? void 0 : typeof opt.notBundle !== "boolean" ? opt.notBundle : extractExternalDeps(context.packageJson)
				});
				return [opt.name, mergeConfig({
					consumer: "server",
					...defaultConfig
				}, opt.options ?? {})];
			}))
		};
	};
	const buildElectronEnvironments = async (builder, environmentOptions) => {
		for (const { name } of environmentOptions) {
			const env = builder.environments[name];
			if (env && !env.isBuilt) await builder.build(env);
		}
	};
	return createElectronPlugin({
		prefix: PLUGIN_PREFIX,
		async dev(pluginContext, server) {
			const { environmentOptions, defaultEnvs } = await resolveOptions(true, server.config.root);
			if (environmentOptions.length === 0) return;
			let builtCount = 0;
			for (const envCfg of Object.values(defaultEnvs)) {
				envCfg.build ??= {};
				envCfg.build.watch ??= {};
				envCfg.build.minify ??= false;
			}
			await buildElectronEnvironments(await createBuilder(withExternalBuiltins({
				configFile: false,
				publicDir: false,
				mode: server.config.mode,
				root: server.config.root,
				envDir: server.config.envDir,
				envPrefix: server.config.envPrefix,
				environments: defaultEnvs,
				plugins: environmentOptions.flatMap((opt) => {
					const name = opt.name;
					return perEnvironmentPlugin(`${PLUGIN_PREFIX}:startup:${name}`, (ctx) => {
						if (ctx.name !== name) return false;
						return {
							name: `${PLUGIN_PREFIX}:startup-hook:${name}`,
							closeBundle() {
								if (++builtCount < environmentOptions.length) return;
								triggerStartup(pluginContext, server, opt);
							}
						};
					});
				})
			})), environmentOptions);
		},
		async build() {},
		async buildConfig(config) {
			const { environmentOptions, defaultEnvs } = await resolveOptions(false, config.root || process.cwd());
			if (environmentOptions.length === 0) return;
			for (const envCfg of Object.values(defaultEnvs)) {
				envCfg.resolve ??= {};
				envCfg.resolve.noExternal ??= true;
			}
			return withExternalBuiltins({
				environments: defaultEnvs,
				builder: { async buildApp(builder) {
					const prevBuildApp = config.builder?.buildApp;
					if (prevBuildApp) await prevBuildApp.call(this, builder);
					const client = builder.environments.client;
					if (client && !client.isBuilt) await builder.build(client);
					await buildElectronEnvironments(builder, environmentOptions);
				} }
			});
		}
	});
}
function electron(options) {
	return electronPluginFactory(() => options);
}
//#endregion
export { electron as default, electronPluginFactory, electronSimple, simpleOptions };
