// vite-plugin-electron/dist/multi-env.cjs
Object.defineProperties(exports, {
	__esModule: { value: true },
	[Symbol.toStringTag]: { value: "Module" }
});
const require_utils = require("./utils-C-Tpxd3l.cjs");
const require_base = require("./base-BISTNoCX.cjs");
const require_notBundle = require("./notBundle-Yz80SRYg.cjs");
let vite = require("vite");
let local_pkg = require("local-pkg");
//#region src/multi-env.ts
/**
* Helper function to create simple API for {@link electron} like `vite-plugin-electron/simple`.
*/
function simpleOptions(options) {
	return Object.entries(options).map(([name, { options, ...rest }]) => {
		switch (name) {
			case "main": return Object.assign(rest, {
				name,
				options: (0, vite.mergeConfig)({ build: require_utils.compatRollupOptions({ rolldownOptions: { platform: "node" } }) }, options ?? {})
			});
			case "preload": return Object.assign(rest, {
				name,
				onstart: rest.onstart || require_base.defaultPreloadOnstart,
				options: (0, vite.mergeConfig)(require_utils.createDefaultPreloadConfig(require_utils.checkESModule()), options ?? {})
			});
			default: return Object.assign(rest, {
				name,
				options
			});
		}
	});
}
function electronSimple(options) {
	return electron(simpleOptions(options));
}
const PLUGIN_PREFIX = "vite-plugin-electron-multi-env";
function electronPluginFactory(options) {
	const resolveOptions = async (isDev, root) => {
		const context = {
			root,
			packageJson: (0, local_pkg.loadPackageJSONSync)(root),
			isDev
		};
		if (!context.packageJson) throw new Error("[vite-plugin-electron] Cannot find package.json");
		const rawOptions = await options(context);
		const envNames = /* @__PURE__ */ new Set();
		const environmentOptions = require_utils.toArray(rawOptions).map((opt, i) => {
			const optionName = opt.name || i;
			const name = `electron_${optionName}`;
			if (envNames.has(name)) throw new Error(`[vite-plugin-electron] Duplicate environment name "${optionName}". Please provide unique "name" properties for each environment in the options array.`);
			envNames.add(name);
			return Object.assign({}, opt, { name });
		});
		return {
			environmentOptions,
			defaultEnvs: Object.fromEntries(environmentOptions.map((opt) => {
				const defaultConfig = require_utils.createElectronViteDefaults(context.packageJson.type === "module", {
					input: opt.input,
					plugins: opt.plugins,
					external: !opt.notBundle ? void 0 : typeof opt.notBundle !== "boolean" ? opt.notBundle : require_notBundle.extractExternalDeps(context.packageJson)
				});
				return [opt.name, (0, vite.mergeConfig)({
					consumer: "server",
					...defaultConfig
				}, opt.options ?? {})];
			}))
		};
	};
	const buildElectronEnvironments = async (builder, environmentOptions) => {
		for (const { name } of environmentOptions) {
			const env = builder.environments[name];
			if (env && !env.isBuilt) await builder.build(env);
		}
	};
	return require_base.createElectronPlugin({
		prefix: PLUGIN_PREFIX,
		async dev(pluginContext, server) {
			const { environmentOptions, defaultEnvs } = await resolveOptions(true, server.config.root);
			if (environmentOptions.length === 0) return;
			let builtCount = 0;
			for (const envCfg of Object.values(defaultEnvs)) {
				envCfg.build ??= {};
				envCfg.build.watch ??= {};
				envCfg.build.minify ??= false;
			}
			await buildElectronEnvironments(await (0, vite.createBuilder)(require_utils.withExternalBuiltins({
				configFile: false,
				publicDir: false,
				mode: server.config.mode,
				root: server.config.root,
				envDir: server.config.envDir,
				envPrefix: server.config.envPrefix,
				environments: defaultEnvs,
				plugins: environmentOptions.flatMap((opt) => {
					const name = opt.name;
					return (0, vite.perEnvironmentPlugin)(`${PLUGIN_PREFIX}:startup:${name}`, (ctx) => {
						if (ctx.name !== name) return false;
						return {
							name: `${PLUGIN_PREFIX}:startup-hook:${name}`,
							closeBundle() {
								if (++builtCount < environmentOptions.length) return;
								require_base.triggerStartup(pluginContext, server, opt);
							}
						};
					});
				})
			})), environmentOptions);
		},
		async build() {},
		async buildConfig(config) {
			const { environmentOptions, defaultEnvs } = await resolveOptions(false, config.root || process.cwd());
			if (environmentOptions.length === 0) return;
			for (const envCfg of Object.values(defaultEnvs)) {
				envCfg.resolve ??= {};
				envCfg.resolve.noExternal ??= true;
			}
			return require_utils.withExternalBuiltins({
				environments: defaultEnvs,
				builder: { async buildApp(builder) {
					const prevBuildApp = config.builder?.buildApp;
					if (prevBuildApp) await prevBuildApp.call(this, builder);
					const client = builder.environments.client;
					if (client && !client.isBuilt) await builder.build(client);
					await buildElectronEnvironments(builder, environmentOptions);
				} }
			});
		}
	});
}
function electron(options) {
	return electronPluginFactory(() => options);
}
//#endregion
exports.default = electron;
exports.electronPluginFactory = electronPluginFactory;
exports.electronSimple = electronSimple;
exports.simpleOptions = simpleOptions;
