// vite-plugin-electron/dist/notBundle-Yz80SRYg.cjs
const require_utils = require("./utils-C-Tpxd3l.cjs");
let local_pkg = require("local-pkg");
//#region src/plugin/notBundle.ts
/**
* Default behavior of extracting external dependencies from package.json.
* - During development, `dependencies`, `devDependencies`, `peerDependencies`, and `optionalDependencies` are externalized.
* - During production, only `dependencies` are externalized.
* @param pkg package.json 's content
*/
function extractExternalDeps(pkg) {
	return Object.keys(require_utils.getIsViteDev() ? {
		...pkg.dependencies,
		...pkg.devDependencies,
		...pkg.peerDependencies,
		...pkg.optionalDependencies
	} : { ...pkg.dependencies });
}
/**
* @see https://github.com/vitejs/vite/blob/v4.4.7/packages/vite/src/node/utils.ts#L140
*/
const bareImportRE = /^(?![a-zA-Z]:)[\w@](?!.*:\/\/)/;
/**
* During dev, we exclude the `cjs` npm-pkg from bundle, mush like Vite :)
*/
function notBundle(options = {}) {
	return {
		name: "vite-plugin-electron:not-bundle",
		enforce: "pre",
		config(cfg) {
			let external;
			if (!options.filter) {
				const pkg = (0, local_pkg.loadPackageJSONSync)(cfg.root);
				if (pkg) external = extractExternalDeps(pkg);
				else {
					console.warn("[vite-plugin-electron:not-bundle] No package.json found in the project root and no filter option provided. All dependencies will be bundled.");
					return;
				}
			} else external = options.filter;
			return { build: require_utils.compatRollupOptions({ rolldownOptions: { external } }) };
		}
	};
}
//#endregion
Object.defineProperty(exports, "bareImportRE", {
	enumerable: true,
	get: function() {
		return bareImportRE;
	}
});
Object.defineProperty(exports, "extractExternalDeps", {
	enumerable: true,
	get: function() {
		return extractExternalDeps;
	}
});
Object.defineProperty(exports, "notBundle", {
	enumerable: true,
	get: function() {
		return notBundle;
	}
});
