// vite-plugin-electron/dist/notBundle-JWunStGa.mjs
import { a as getIsViteDev, n as compatRollupOptions } from "./utils-DzpHSY0b.mjs";
import { loadPackageJSONSync } from "local-pkg";
//#region src/plugin/notBundle.ts
/**
* Default behavior of extracting external dependencies from package.json.
* - During development, `dependencies`, `devDependencies`, `peerDependencies`, and `optionalDependencies` are externalized.
* - During production, only `dependencies` are externalized.
* @param pkg package.json 's content
*/
function extractExternalDeps(pkg) {
	return Object.keys(getIsViteDev() ? {
		...pkg.dependencies,
		...pkg.devDependencies,
		...pkg.peerDependencies,
		...pkg.optionalDependencies
	} : { ...pkg.dependencies });
}
/**
* @see https://github.com/vitejs/vite/blob/v4.4.7/packages/vite/src/node/utils.ts#L140
*/
const bareImportRE = /^(?![a-zA-Z]:)[\w@](?!.*:\/\/)/;
/**
* During dev, we exclude the `cjs` npm-pkg from bundle, mush like Vite :)
*/
function notBundle(options = {}) {
	return {
		name: "vite-plugin-electron:not-bundle",
		enforce: "pre",
		config(cfg) {
			let external;
			if (!options.filter) {
				const pkg = loadPackageJSONSync(cfg.root);
				if (pkg) external = extractExternalDeps(pkg);
				else {
					console.warn("[vite-plugin-electron:not-bundle] No package.json found in the project root and no filter option provided. All dependencies will be bundled.");
					return;
				}
			} else external = options.filter;
			return { build: compatRollupOptions({ rolldownOptions: { external } }) };
		}
	};
}
//#endregion
export { extractExternalDeps as n, notBundle as r, bareImportRE as t };
