// vite-plugin-electron/dist/utils-C-Tpxd3l.cjs
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let vite = require("vite");
let node_path = require("node:path");
node_path = __toESM(node_path, 1);
let node_fs = require("node:fs");
node_fs = __toESM(node_fs, 1);
let node_module = require("node:module");
let local_pkg = require("local-pkg");
//#region src/utils.ts
function toArray(item) {
	if (item === void 0) return [];
	return Array.isArray(item) ? item : [item];
}
function resolveBuiltinExternals(external) {
	const nodeProtocolRE = /^node:/;
	const builtins = [
		"electron",
		"electron/main",
		"electron/renderer",
		"electron/common",
		"electron/utility",
		"original-fs",
		...node_module.builtinModules,
		nodeProtocolRE
	];
	if (Array.isArray(external)) return builtins.concat(external);
	if (typeof external === "string" || external instanceof RegExp) return builtins.concat([external]);
	if (typeof external === "function") {
		const original = external;
		return function(source, importer, isResolved) {
			if (builtins.includes(source) || nodeProtocolRE.test(source)) return true;
			return original(source, importer, isResolved);
		};
	}
	return builtins;
}
/**
* Normalize the build config key for the current Vite version.
* Vite 8+ reads `build.rolldownOptions`, while Vite < 8 reads `build.rollupOptions`.
*/
function getBuildOptions(build) {
	if (!build) return;
	return build.rolldownOptions || build.rollupOptions;
}
function ensureBuildOptions(build) {
	const options = getBuildOptions(build) || {};
	if (!build.rolldownOptions && !build.rollupOptions) build.rolldownOptions = options;
	return options;
}
/**
* Keep only the build config key supported by the current Vite version.
* This converts the user-facing config shape between `rolldownOptions` and `rollupOptions`
* without changing the actual option payload.
*/
function setBuildOptions(build, viteVersion = vite.version) {
	if (!build) return;
	const options = getBuildOptions(build);
	if (options) if (Number.parseInt(viteVersion) < 8) {
		delete build.rolldownOptions;
		build.rollupOptions = options;
		for (const output of toArray(options.output)) if (typeof output.codeSplitting === "boolean") output.inlineDynamicImports = !output.codeSplitting;
	} else {
		delete build.rollupOptions;
		build.rolldownOptions = options;
	}
}
function compatRollupOptions(options, viteVersion = vite.version) {
	setBuildOptions(options, viteVersion);
	return options;
}
function checkESModule(root) {
	return (0, local_pkg.loadPackageJSONSync)(root)?.type === "module";
}
function createDefaultPreloadConfig(esmodule, input) {
	const fileExt = esmodule ? "mjs" : "js";
	return { build: compatRollupOptions({ rolldownOptions: {
		input,
		platform: "node",
		output: {
			format: "cjs",
			codeSplitting: false,
			entryFileNames: `[name].${fileExt}`,
			chunkFileNames: `[name].${fileExt}`,
			assetFileNames: "[name].[ext]"
		}
	} }) };
}
function createElectronViteDefaults(esmodule = false, options = {}) {
	return {
		build: compatRollupOptions({
			lib: options.entry ? {
				entry: options.entry,
				formats: esmodule ? ["es"] : ["cjs"],
				fileName: () => "[name].js"
			} : void 0,
			outDir: "dist-electron",
			emptyOutDir: false,
			rolldownOptions: {
				platform: "node",
				...options.input !== void 0 ? {
					input: options.input,
					output: { format: esmodule ? "es" : "cjs" }
				} : {},
				...options.plugins !== void 0 ? { plugins: options.plugins } : {},
				...options.external !== void 0 ? { external: options.external } : {}
			}
		}),
		resolve: {
			conditions: ["node"],
			mainFields: [
				"module",
				"jsnext:main",
				"jsnext"
			]
		},
		define: { "process.env": "process.env" }
	};
}
/** Resolve the default Vite's `InlineConfig` for build Electron-Main */
function resolveViteConfig(options) {
	return resolveViteConfigBase(checkESModule(), options);
}
/** Resolve the default Vite's `InlineConfig` for build Electron-Main */
function resolveViteConfigBase(esmodule, options) {
	const defaultConfig = createElectronViteDefaults(esmodule, { entry: options.entry });
	defaultConfig.configFile = false;
	defaultConfig.publicDir = false;
	return (0, vite.mergeConfig)(defaultConfig, options?.vite || {});
}
/**
* Externalize Electron and Node builtins in both build modes.
* The build config keeps `rolldownOptions` on Vite 8+ and `rollupOptions` on Vite < 8.
*/
function withExternalBuiltins(config) {
	config.build ??= {};
	const buildOptions = ensureBuildOptions(config.build);
	buildOptions.external = resolveBuiltinExternals(buildOptions.external);
	setBuildOptions(config.build);
	if (config.environments) for (const environment of Object.values(config.environments)) {
		environment.build ??= {};
		const environmentBuildOptions = ensureBuildOptions(environment.build);
		environmentBuildOptions.external = resolveBuiltinExternals(environmentBuildOptions.external);
		setBuildOptions(environment.build);
	}
	return config;
}
/**
* @see https://github.com/vitejs/vite/blob/v4.0.1/packages/vite/src/node/constants.ts#L137-L147
*/
function resolveHostname(hostname) {
	const loopbackHosts = new Set([
		"localhost",
		"127.0.0.1",
		"::1",
		"0000:0000:0000:0000:0000:0000:0000:0001"
	]);
	const wildcardHosts = new Set([
		"0.0.0.0",
		"::",
		"0000:0000:0000:0000:0000:0000:0000:0000"
	]);
	return loopbackHosts.has(hostname) || wildcardHosts.has(hostname) ? "localhost" : hostname;
}
function resolveServerUrl(server) {
	const addressInfo = server.httpServer?.address();
	const isAddressInfo = (x) => x?.address;
	if (isAddressInfo(addressInfo)) {
		const { address, port } = addressInfo;
		const hostname = resolveHostname(address);
		const options = server.config.server;
		const protocol = options.https ? "https" : "http";
		const devBase = server.config.base;
		const path = typeof options.open === "string" ? options.open : devBase;
		return path.startsWith("http") ? path : `${protocol}://${hostname}:${port}${path}`;
	}
}
/** @see https://github.com/vitejs/vite/blob/v5.4.9/packages/vite/src/node/build.ts#L489-L504 */
function resolveInput(config) {
	const options = config.build;
	const { root } = config;
	const libOptions = options.lib;
	const buildOptions = getBuildOptions(options);
	const resolve = (p) => node_path.default.resolve(root, p);
	const normalizeInput = (input) => {
		if (typeof input === "string") return resolve(input);
		if (Array.isArray(input)) return input.map((file) => resolve(file));
		if (input && typeof input === "object") return Object.fromEntries(Object.entries(input).map(([alias, file]) => [alias, resolve(file)]));
		return input;
	};
	const input = libOptions ? normalizeInput(buildOptions?.input) || (typeof libOptions.entry === "string" ? resolve(libOptions.entry) : Array.isArray(libOptions.entry) ? libOptions.entry.map(resolve) : Object.fromEntries(Object.entries(libOptions.entry).map(([alias, file]) => [alias, resolve(file)]))) : normalizeInput(buildOptions?.input);
	if (input) return input;
	const indexHtml = resolve("index.html");
	return node_fs.default.existsSync(indexHtml) ? indexHtml : void 0;
}
const MOCK_INDEX_HTML = `<!doctype html>
<html lang="en">
  <head>
    <title>vite-plugin-electron</title>
  </head>
  <body>
    <div>An entry file for electron renderer process.</div>
  </body>
</html>`;
/**
* Write a temporary mock `index.html` at `filepath` so that Vite has a valid
* entry point when no real `index.html` and no configured input exist.
*
* Returns an async cleanup function.  Before removing the file it re-reads the
* content and skips deletion when another plugin or tool has replaced the file
* in the meantime, guarding against accidental removal of real HTML written in
* parallel.
*/
function setupMockHtml(config, isBuild, logger) {
	const { root, build: buildConfig } = config;
	const mockFilepath = node_path.default.join(root, "index.html");
	const distFilepath = node_path.default.resolve(root, buildConfig.outDir, "index.html");
	logger.info(`[vite-plugin-electron] No entry found, writing mock ${mockFilepath}`);
	node_fs.default.writeFileSync(mockFilepath, MOCK_INDEX_HTML, "utf-8");
	return async () => {
		if (await node_fs.default.promises.readFile(mockFilepath, "utf-8").catch(() => null) === MOCK_INDEX_HTML) await node_fs.default.promises.unlink(mockFilepath).catch((err) => {
			logger.warn(`[vite-plugin-electron] Failed to remove mock ${mockFilepath}:`, err);
		});
		if (isBuild && distFilepath) await node_fs.default.promises.unlink(distFilepath).catch(() => {});
	};
}
const VITE_IS_DEV = "VITE_IS_DEV";
function setIsViteDev() {
	process.env[VITE_IS_DEV] = "true";
}
function getIsViteDev() {
	return ["1", "true"].includes((process.env[VITE_IS_DEV] ?? "").trim().toLowerCase());
}
//#endregion
Object.defineProperty(exports, "__toESM", {
	enumerable: true,
	get: function() {
		return __toESM;
	}
});
Object.defineProperty(exports, "checkESModule", {
	enumerable: true,
	get: function() {
		return checkESModule;
	}
});
Object.defineProperty(exports, "compatRollupOptions", {
	enumerable: true,
	get: function() {
		return compatRollupOptions;
	}
});
Object.defineProperty(exports, "createDefaultPreloadConfig", {
	enumerable: true,
	get: function() {
		return createDefaultPreloadConfig;
	}
});
Object.defineProperty(exports, "createElectronViteDefaults", {
	enumerable: true,
	get: function() {
		return createElectronViteDefaults;
	}
});
Object.defineProperty(exports, "getIsViteDev", {
	enumerable: true,
	get: function() {
		return getIsViteDev;
	}
});
Object.defineProperty(exports, "resolveInput", {
	enumerable: true,
	get: function() {
		return resolveInput;
	}
});
Object.defineProperty(exports, "resolveServerUrl", {
	enumerable: true,
	get: function() {
		return resolveServerUrl;
	}
});
Object.defineProperty(exports, "resolveViteConfig", {
	enumerable: true,
	get: function() {
		return resolveViteConfig;
	}
});
Object.defineProperty(exports, "resolveViteConfigBase", {
	enumerable: true,
	get: function() {
		return resolveViteConfigBase;
	}
});
Object.defineProperty(exports, "setIsViteDev", {
	enumerable: true,
	get: function() {
		return setIsViteDev;
	}
});
Object.defineProperty(exports, "setupMockHtml", {
	enumerable: true,
	get: function() {
		return setupMockHtml;
	}
});
Object.defineProperty(exports, "toArray", {
	enumerable: true,
	get: function() {
		return toArray;
	}
});
Object.defineProperty(exports, "withExternalBuiltins", {
	enumerable: true,
	get: function() {
		return withExternalBuiltins;
	}
});
