// vite-plugin-electron/dist/utils-DzpHSY0b.mjs
import { builtinModules } from "node:module";
import { mergeConfig, version } from "vite";
import path from "node:path";
import fs from "node:fs";
import { loadPackageJSONSync } from "local-pkg";
//#region src/utils.ts
function toArray(item) {
	if (item === void 0) return [];
	return Array.isArray(item) ? item : [item];
}
function resolveBuiltinExternals(external) {
	const nodeProtocolRE = /^node:/;
	const builtins = [
		"electron",
		"electron/main",
		"electron/renderer",
		"electron/common",
		"electron/utility",
		"original-fs",
		...builtinModules,
		nodeProtocolRE
	];
	if (Array.isArray(external)) return builtins.concat(external);
	if (typeof external === "string" || external instanceof RegExp) return builtins.concat([external]);
	if (typeof external === "function") {
		const original = external;
		return function(source, importer, isResolved) {
			if (builtins.includes(source) || nodeProtocolRE.test(source)) return true;
			return original(source, importer, isResolved);
		};
	}
	return builtins;
}
/**
* Normalize the build config key for the current Vite version.
* Vite 8+ reads `build.rolldownOptions`, while Vite < 8 reads `build.rollupOptions`.
*/
function getBuildOptions(build) {
	if (!build) return;
	return build.rolldownOptions || build.rollupOptions;
}
function ensureBuildOptions(build) {
	const options = getBuildOptions(build) || {};
	if (!build.rolldownOptions && !build.rollupOptions) build.rolldownOptions = options;
	return options;
}
/**
* Keep only the build config key supported by the current Vite version.
* This converts the user-facing config shape between `rolldownOptions` and `rollupOptions`
* without changing the actual option payload.
*/
function setBuildOptions(build, viteVersion = version) {
	if (!build) return;
	const options = getBuildOptions(build);
	if (options) if (Number.parseInt(viteVersion) < 8) {
		delete build.rolldownOptions;
		build.rollupOptions = options;
		for (const output of toArray(options.output)) if (typeof output.codeSplitting === "boolean") output.inlineDynamicImports = !output.codeSplitting;
	} else {
		delete build.rollupOptions;
		build.rolldownOptions = options;
	}
}
function compatRollupOptions(options, viteVersion = version) {
	setBuildOptions(options, viteVersion);
	return options;
}
function checkESModule(root) {
	return loadPackageJSONSync(root)?.type === "module";
}
function createDefaultPreloadConfig(esmodule, input) {
	const fileExt = esmodule ? "mjs" : "js";
	return { build: compatRollupOptions({ rolldownOptions: {
		input,
		platform: "node",
		output: {
			format: "cjs",
			codeSplitting: false,
			entryFileNames: `[name].${fileExt}`,
			chunkFileNames: `[name].${fileExt}`,
			assetFileNames: "[name].[ext]"
		}
	} }) };
}
function createElectronViteDefaults(esmodule = false, options = {}) {
	return {
		build: compatRollupOptions({
			lib: options.entry ? {
				entry: options.entry,
				formats: esmodule ? ["es"] : ["cjs"],
				fileName: () => "[name].js"
			} : void 0,
			outDir: "dist-electron",
			emptyOutDir: false,
			rolldownOptions: {
				platform: "node",
				...options.input !== void 0 ? {
					input: options.input,
					output: { format: esmodule ? "es" : "cjs" }
				} : {},
				...options.plugins !== void 0 ? { plugins: options.plugins } : {},
				...options.external !== void 0 ? { external: options.external } : {}
			}
		}),
		resolve: {
			conditions: ["node"],
			mainFields: [
				"module",
				"jsnext:main",
				"jsnext"
			]
		},
		define: { "process.env": "process.env" }
	};
}
/** Resolve the default Vite's `InlineConfig` for build Electron-Main */
function resolveViteConfig(options) {
	return resolveViteConfigBase(checkESModule(), options);
}
/** Resolve the default Vite's `InlineConfig` for build Electron-Main */
function resolveViteConfigBase(esmodule, options) {
	const defaultConfig = createElectronViteDefaults(esmodule, { entry: options.entry });
	defaultConfig.configFile = false;
	defaultConfig.publicDir = false;
	return mergeConfig(defaultConfig, options?.vite || {});
}
/**
* Externalize Electron and Node builtins in both build modes.
* The build config keeps `rolldownOptions` on Vite 8+ and `rollupOptions` on Vite < 8.
*/
function withExternalBuiltins(config) {
	config.build ??= {};
	const buildOptions = ensureBuildOptions(config.build);
	buildOptions.external = resolveBuiltinExternals(buildOptions.external);
	setBuildOptions(config.build);
	if (config.environments) for (const environment of Object.values(config.environments)) {
		environment.build ??= {};
		const environmentBuildOptions = ensureBuildOptions(environment.build);
		environmentBuildOptions.external = resolveBuiltinExternals(environmentBuildOptions.external);
		setBuildOptions(environment.build);
	}
	return config;
}
/**
* @see https://github.com/vitejs/vite/blob/v4.0.1/packages/vite/src/node/constants.ts#L137-L147
*/
function resolveHostname(hostname) {
	const loopbackHosts = new Set([
		"localhost",
		"127.0.0.1",
		"::1",
		"0000:0000:0000:0000:0000:0000:0000:0001"
	]);
	const wildcardHosts = new Set([
		"0.0.0.0",
		"::",
		"0000:0000:0000:0000:0000:0000:0000:0000"
	]);
	return loopbackHosts.has(hostname) || wildcardHosts.has(hostname) ? "localhost" : hostname;
}
function resolveServerUrl(server) {
	const addressInfo = server.httpServer?.address();
	const isAddressInfo = (x) => x?.address;
	if (isAddressInfo(addressInfo)) {
		const { address, port } = addressInfo;
		const hostname = resolveHostname(address);
		const options = server.config.server;
		const protocol = options.https ? "https" : "http";
		const devBase = server.config.base;
		const path = typeof options.open === "string" ? options.open : devBase;
		return path.startsWith("http") ? path : `${protocol}://${hostname}:${port}${path}`;
	}
}
/** @see https://github.com/vitejs/vite/blob/v5.4.9/packages/vite/src/node/build.ts#L489-L504 */
function resolveInput(config) {
	const options = config.build;
	const { root } = config;
	const libOptions = options.lib;
	const buildOptions = getBuildOptions(options);
	const resolve = (p) => path.resolve(root, p);
	const normalizeInput = (input) => {
		if (typeof input === "string") return resolve(input);
		if (Array.isArray(input)) return input.map((file) => resolve(file));
		if (input && typeof input === "object") return Object.fromEntries(Object.entries(input).map(([alias, file]) => [alias, resolve(file)]));
		return input;
	};
	const input = libOptions ? normalizeInput(buildOptions?.input) || (typeof libOptions.entry === "string" ? resolve(libOptions.entry) : Array.isArray(libOptions.entry) ? libOptions.entry.map(resolve) : Object.fromEntries(Object.entries(libOptions.entry).map(([alias, file]) => [alias, resolve(file)]))) : normalizeInput(buildOptions?.input);
	if (input) return input;
	const indexHtml = resolve("index.html");
	return fs.existsSync(indexHtml) ? indexHtml : void 0;
}
const MOCK_INDEX_HTML = `<!doctype html>
<html lang="en">
  <head>
    <title>vite-plugin-electron</title>
  </head>
  <body>
    <div>An entry file for electron renderer process.</div>
  </body>
</html>`;
/**
* Write a temporary mock `index.html` at `filepath` so that Vite has a valid
* entry point when no real `index.html` and no configured input exist.
*
* Returns an async cleanup function.  Before removing the file it re-reads the
* content and skips deletion when another plugin or tool has replaced the file
* in the meantime, guarding against accidental removal of real HTML written in
* parallel.
*/
function setupMockHtml(config, isBuild, logger) {
	const { root, build: buildConfig } = config;
	const mockFilepath = path.join(root, "index.html");
	const distFilepath = path.resolve(root, buildConfig.outDir, "index.html");
	logger.info(`[vite-plugin-electron] No entry found, writing mock ${mockFilepath}`);
	fs.writeFileSync(mockFilepath, MOCK_INDEX_HTML, "utf-8");
	return async () => {
		if (await fs.promises.readFile(mockFilepath, "utf-8").catch(() => null) === MOCK_INDEX_HTML) await fs.promises.unlink(mockFilepath).catch((err) => {
			logger.warn(`[vite-plugin-electron] Failed to remove mock ${mockFilepath}:`, err);
		});
		if (isBuild && distFilepath) await fs.promises.unlink(distFilepath).catch(() => {});
	};
}
const VITE_IS_DEV = "VITE_IS_DEV";
function setIsViteDev() {
	process.env[VITE_IS_DEV] = "true";
}
function getIsViteDev() {
	return ["1", "true"].includes((process.env[VITE_IS_DEV] ?? "").trim().toLowerCase());
}
//#endregion
export { getIsViteDev as a, resolveViteConfig as c, setupMockHtml as d, toArray as f, createElectronViteDefaults as i, resolveViteConfigBase as l, compatRollupOptions as n, resolveInput as o, withExternalBuiltins as p, createDefaultPreloadConfig as r, resolveServerUrl as s, checkESModule as t, setIsViteDev as u };
